rfd/0050-join-methods.md

```diff
@@ -0,0 +1,117 @@
+---
+authors: Nic Klaassen <nic@goteleport.com>
+state: Draft
+---
+
+# RFD 50 - Cluster Join Methods and Endpoints
+
+## What
+
+There are multiple ways a node can get credentials for a Teleport cluster. This
+document describes the current node joining methods and details upcoming changes
+for the IAM join method (RFD 41) and Certificate Bot
+(https://github.com/gravitational/teleport/pull/7986).
+
+## Why
+
+This RFD can serve as a reference for all of the ways Teleport nodes can get
+cluster credentials in order to facilitate discussion of the current methods and
+new endpoints that will be added for the IAM join method and Certificate Bot.
+
+## Details
+
+When "joining" a cluster, a Teleport node must authenticate itself using either
+a secret token or one of the new AWS join methods (EC2 or IAM method). If the
+node can successfully authenticate, the Auth server will return signed SSH and
+TLS certificates to the node for the requested role (Proxy, Node, Kubernetes,
+etc). The token or AWS join method is only used once for the first-time connect
+and from that point on signed certificates are used.
+
+The joining node also needs to be able to trust that the Teleport cluster is
+authentic. It can do this by using CA pins configured on the node, in the case
+of IoT nodes connecting to the proxy, the proxy TLS cert can be trusted through
+PKI.
+
+### Current Join Endpoints
+
+#### Auth `POST /tokens/register`
+Join methods: token, EC2
+Server trust: CA pins
+This is the main token register endpoint on the Auth server. Accepts tokens or
+EC2 Identity Documents to authenticate the client.
+
+#### Auth `POST /tokens/register/auth`
+Join Methods: token
+Server trust: CA pins
+Looks like this is a legacy endpoint the just deletes the given token. Only
+works for tokens with the `Auth` role. I can't find anything that calls this, so
+it can probably be scheduled for deletion.
+
+#### Auth `rpc GenerateHostCerts(HostCertsRequest) returns (Certs)`
+Client trust: mTLS
+Server trust: mTLS
+This is an mTLS authenticated gRPC endpoint for cert renewal. The node must
+first join the cluster using another endpoint to get its first client
+certificate used for the mTLS connection.
+
+#### Proxy `POST /webapi/host/credentials`
+Join methods: token, EC2
+Server trust: PKI
+This is the proxy endpoint for registering IoT nodes that don't have a direct connection to the auth server, it basically forwards to
+the Auth `/tokens/register` endpoint.
+
+### IAM Join Method
+The IAM join method (introduced in RFD 41) requires gRPC methods rather than
```

Comment:
RFD 41 says that we chose the "EC2 method", is that still correct?

Reply:
Thanks for pointing that out. The EC2 method was chosen to implement first, we are now implementing the IAM method as well (which will support Cloud). Edited RFD 41.